apps/apps.c

```diff
@@ -2851,3 +2970,164 @@ void make_uppercase(char *string)
+static int tempfile_gets(BIO *b, char *buf, int size);
+static int tempfile_puts(BIO *b, const char *str);
+
+static const BIO_METHOD *BIO_f_tempfile(void)
+{
+    if (methods_tempfile == NULL) {
+        methods_tempfile = BIO_meth_new(BIO_TYPE_TEMP_FILE,
+                                        "Temporary file filter");
+        if (methods_tempfile == NULL
+            || !BIO_meth_set_write(methods_tempfile, tempfile_write)
+            || !BIO_meth_set_read(methods_tempfile, tempfile_read)
+            || !BIO_meth_set_puts(methods_tempfile, tempfile_puts)
+            || !BIO_meth_set_gets(methods_tempfile, tempfile_gets)
+            || !BIO_meth_set_ctrl(methods_tempfile, tempfile_ctrl)
+            || !BIO_meth_set_create(methods_tempfile, tempfile_new)
+            || !BIO_meth_set_destroy(methods_tempfile, tempfile_free))
+            return NULL;
+    }
+    return methods_tempfile;
+}
+
+void bio_tempfile_cleanup(void)
+{
+    BIO_meth_free(methods_tempfile);
+}
+
+static int tempfile_new(BIO *b)
+{
+    TEMPFILE_CTX *temp;
+
+    temp = app_malloc(sizeof(*temp), "temp file");
+    temp->filename = NULL;
+    temp->tmp_filename = NULL;
+    BIO_set_data(b, temp);
+    BIO_set_init(b, 1);
+
+    return 1;
+}
+
+static int tempfile_free(BIO *b)
+{
+    TEMPFILE_CTX *temp;
+
+    if (b == NULL)
+        return 0;
+
+    temp = BIO_get_data(b);
+    if (temp->filename != NULL && temp->tmp_filename != NULL) {
+#if !defined(_WIN32) && !defined(__VMS)
+        rename(temp->tmp_filename, temp->filename);
+#endif
+    }
+    if (temp->filename)
+        OPENSSL_free(temp->filename);
+    if (temp->tmp_filename)
+        OPENSSL_free(temp->tmp_filename);
+    OPENSSL_free(temp);
+    BIO_set_data(b, NULL);
+    BIO_set_init(b, 0);
+
+    return 1;
+}
+
+static int tempfile_read(BIO *b, char *out, int outl)
+{
+    int ret = 0;
+    BIO *next = BIO_next(b);
+
+    if (out == NULL || outl == 0)
+        return 0;
+    if (next == NULL)
+        return 0;
+
+    ret = BIO_read(next, out, outl);
+    return ret;
+}
+
+static int tempfile_write(BIO *b, const char *in, int inl)
+{
+    BIO *next = BIO_next(b);
+
+    if ((in == NULL) || (inl <= 0))
+        return 0;
+    if (next == NULL)
+        return 0;
+
+    return BIO_write(next, in, inl);
+}
+
+static long tempfile_ctrl(BIO *b, int cmd, long num, void *ptr)
+{
+    long ret = 1;
+    BIO *next = BIO_next(b);
+    TEMPFILE_CTX *temp;
+    char **p;
+
+    if (next == NULL)
+        return 0;
+    switch (cmd) {
+    case BIO_C_SET_FILENAME:
+        temp = BIO_get_data(b);
+        if (num == 1)
+            p = &temp->filename;
+        else
+            p = &temp->tmp_filename;
+        if (*p)
+            OPENSSL_free(*p);
+        *p = OPENSSL_strdup((char *)ptr);
+        if (p == NULL)
+            return 0;
+        break;
+    default:
+        ret = BIO_ctrl(next, cmd, num, ptr);
+        break;
+    }
+    return ret;
+}
+
+static int tempfile_gets(BIO *b, char *buf, int size)
+{
```

Comment:
Would it make sense to pass this to the next in the chain rather than implementing it ourselves?

Reply:
Yes